contracts/QuestTerminalKey.sol

```diff
@@ -0,0 +1,192 @@
+    struct Discount {
+        uint16 percentage; //in BIPS
+        uint16 maxUses;
+        uint16 usedCount;
+    }
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    function initialize(
+        address royaltyRecipient_,
+        address minterAddress_,
+        address questFactoryAddress_,
+        uint royaltyFee_,
+        address owner_
+    ) external initializer {
+        __ERC721_init('QuestTerminalKey', 'QTK');
+        __ERC721URIStorage_init();
+        __Ownable_init(owner_);
+        royaltyRecipient = royaltyRecipient_;
+        minterAddress = minterAddress_;
+        questFactoryAddress = questFactoryAddress_;
+        royaltyFee = royaltyFee_;
+    }
+
+    modifier onlyMinter() {
+        require(msg.sender == minterAddress, 'Only minter');
+        _;
+    }
+
+    modifier onlyQuestFactory() {
+        require(msg.sender == questFactoryAddress, 'Only quest factory');
+        _;
+    }
+
+    function _baseURI() internal pure override returns (string memory) {
+        return "https://api.rabbithole.gg/nft/qtk/";
+    }
+
+    /// @dev modifier to check for zero address
+    /// @param _address the address to check
+    modifier nonZeroAddress(address _address) {
+        require(_address != address(0), 'Zero address');
+        _;
+    }
+
+    /// @dev set the royalty recipient
+    /// @param royaltyRecipient_ the address of the royalty recipient
+    function setRoyaltyRecipient(address royaltyRecipient_) external nonZeroAddress(royaltyRecipient_) onlyOwner {
+        royaltyRecipient = royaltyRecipient_;
+    }
+
+    /// @dev set the minter address
+    /// @param minterAddress_ the address of the minter
+    function setMinterAddress(address minterAddress_) external nonZeroAddress(minterAddress_) onlyOwner {
+        minterAddress = minterAddress_;
+        emit MinterAddressSet(minterAddress_);
+    }
+
+    /// @dev set the quest factory address
+    /// @param questFactoryAddress_ the address of the quest factory
+    function setQuestFactoryAddress(address questFactoryAddress_) external nonZeroAddress(questFactoryAddress_) onlyOwner {
+        questFactoryAddress = questFactoryAddress_;
+        emit QuestFactoryAddressSet(questFactoryAddress_);
+    }
+
+    /// @dev set the royalty fee
+    /// @param royaltyFee_ the royalty fee
+    function setRoyaltyFee(uint256 royaltyFee_) external onlyOwner {
+        royaltyFee = royaltyFee_;
+        emit RoyaltyFeeSet(royaltyFee_);
+    }
+
+    /// @dev mint a QuestTerminalKey NFT
+    /// @param to_ the address to mint to
+    /// @param discountPercentage_ the discount percentage
+    /// @param maxDiscountUses_ the max discount uses
+    function mint(address to_, uint16 discountPercentage_, uint16 maxDiscountUses_) external onlyMinter {
+        require(discountPercentage_ <= 10000, 'Invalid discount percentage');
+        require(maxDiscountUses_ <= 1000, 'Invalid max discount uses');
```

Comment:
So the minimum amount of times the nft can use the discount is 1001? Is there a reason for that?

Reply:
Oops sorry read this backwards LOL. It requires 1000 or less

I guess we want to set an upper limit for the amount of times the discount can be used